### DISCUSSION TITLE
Tracing directly to queue instead of http api

### DISCUSSION THREAD

**Discussion by exotikh3** [langfuse/langfuse#2983] [Support] [answered]
Dear Langfuse maintainers,

First of all thanks for the great tool.

We are early adopters of Langfuse and looking to start with self-hosted option. Yet our usecase will likely require some customization. 
1. we would really like to avoid http calls from the place where LLMs are called and send traces and other information to some queue.
2. Read from this queue and ingest into Postgres DB by an independent worker and store it in a tenanted schema.
3. Bring up Langfuse and point it to the specific tenant schema.

Of course this will require additional customization of Python SDK for step 1 + writing our own ingestion worker + changing some Langfuse code to be able to support tenanted schema. What concerns me in this setup is that as Langfuse 3.0 coming default ingestion workers that are coming with Langfuse would have a lot of logic that will make having our own worker that are ingesting into database nearly useless.

So the question is whether it actually adequate to replicate ingestion into database in our own worker or it would make no sense after Langfuse 3.0 will appear?

**Comment by marcklingen** [comment]
Thanks for sharing your thoughts!

Langfuse v3 will include an ingestion api which writes to a queue to acknowledge newly ingested traces quickly and handle spikes in traffic efficiently. From this queue the worker container will flush events to the database in batches.

I would not create/maintain this logic myself if I was you, what mainly drives the need to add this logic right now? The SDKs make all requests asynchronously/non-blocking which thus should not affect your main application. If you have very fluctuating traffic, you could scale your db to cope with the traffic (usually constrained by IOPS).

**Reply by exotikh3** [reply]
The idea is mainly motivated by the constraints:

1. We do LLM related processing inside of AWS Lambdas and would like to stick with it for a while. So we will need to do a flush after every Lambda execution. And would really prefer to flush into some queue rather than into http service.
2. We do store customer data in tenanted setup. And likely in future we will add some privacy/security features related to such separation. Therefore it's very questionable whether or not we could store prompts from different customers in one database. Spawn at least 1 Langfuse worker per customer looks like infra overhead for us at the current moment. Not storing input/outputs is killing the idea of using tracing. And tracing is a very important part of Langfuse functionality for us.

And as there is no real alternative for the Langfuse we can think only about customization. By the way is there any target month for Langfuse 3.0 to appear for early testing? Or maybe there is a branch I could look into to have a better sense of what is coming?

**Reply by marcklingen** [reply, marked as answer]
Hi @exotikh3 

Fully understand your use case

> We do LLM related processing inside of AWS Lambdas and would like to stick with it for a while. So we will need to do a flush after every Lambda execution. And would really prefer to flush into some queue rather than into http service.

This will not be officially supported by this project as it will not be a viable option for most users who need to use http. If the http service acknowledges and writes to a queue, it is very scalable. What is your concern here?

> We do store customer data in tenanted setup. And likely in future we will add some privacy/security features related to such separation. Therefore it's very questionable whether or not we could store prompts from different customers in one database. Spawn at least 1 Langfuse worker per customer looks like infra overhead for us at the current moment. Not storing input/outputs is killing the idea of using tracing. And tracing is a very important part of Langfuse functionality for us.

You could log each user to their own project to logically separate data in Langfuse, however, this seems to be a bit overkill. Do you already know what the hard requirement here is?

> And as there is no real alternative for the Langfuse we can think only about customization. By the way is there any target month for Langfuse 3.0 to appear for early testing? Or maybe there is a branch I could look into to have a better sense of what is coming?

Aiming for Sep-earlyNov, we will reach out via the mailing list once a preview is available.

**Reply by exotikh3** [reply]
@marcklingen thanks for the response.

> If the http service acknowledges and writes to a queue, it is very scalable. What is your concern here?

As for internal vs external queue I would argue that external one is more reliable, easier to observe and manage. It is also more transparent for non-langfuse users (like SRE and DevOps teams who are not aware about Langfuse internals). Like the basic scenario I could come up with is what is going to happen if writing to DB is going to start lagging? Will Langfuse will be able to scale quickly enough in order for workers not to go down? Or will http endpoint stop receiving requests when the queue is full in order not to go down with OOM error? In case with external queue it is also a valid point but usually external queue is better monitored -> faster proper reaction will appear. This will appear mostly in high load environment but I think it is what you are looking forward to if you decided to introduce Clickhouse as a DB.

Anyway. it is just me being technical nerd. Being you I would have chosen the same or similar approach with http server because it allows broader amount of people to use Langfuse. Writing connectors for all the queues out there is extremely time consuming and probably not worth it. But being me if I can do relatively easy modification with external queue - I would have gone for it. 

> You could log each user to their own project to logically separate data in Langfuse, however, this seems to be a bit overkill. Do you already know what the hard requirement here is?

Hard requirement is that we should isolate customers data as much as possible. Example of the usecase: client starts to use our platform and then wants to run it on their private cloud. It could be a real scenario for big enterprise companies. And we need to migrate the data. And customer will be very strict about us not having their data after migration. Some other wild examples might occur but that's the world of enterprise software as I see it.

Thanks again for your answer and for Langfuse.

P.S. if you are interested: We will follow your advice about not modifying Langfuse code. But we will put writing traces for different customers under feature flag and will set up a very strict data retaining policy. So F.E. we will be able to have only one customer data in DB for very limited amount of time.

**Reply by marcklingen** [reply]
Thanks for sharing the additional context, this is super interesting!

You can migrate data via the api or python/js sdks if you want to move data between instances while relying on a stable interface without touching the db itself.
